File: Greedy/3474. Lexicographically Smallest Generated String/solution.py
class Solution:
    def generateString(self, str1: str, str2: str) -> str:
        """
        Greedy
        """
        n, m = len(str1), len(str2)
        word = [""] * (n + m - 1)
        T = [False] * (n + m - 1)

        # 1. fill all the `T`s
        for i in range(n):
            if str1[i] == "T":
                for j in range(m):
                    if word[i + j] and word[i + j] != str2[j]:
                        return ""
                    word[i + j] = str2[j]
                    T[i + j] = True

        # 2. fill all the `F`s with 'a' first
        for i in range(n + m - 1):
            if not word[i]:
                word[i] = "a"

        for i in range(n):
            if str1[i] == "F":
                # 3. check if aleady different
                diff = False
                for j in range(m):
                    if word[i + j] != str2[j]:
                        diff = True
                        break
                if diff:
                    continue

                # 4. find a char to change from the end to the begin
                change = False
                for j in range(m - 1, -1, -1):
                    # if cur is `T`, it cannot be changed, continue
                    if T[i + j]: continue
                    
                    if word[i+j] == "a":
                        word[i + j] = "b"

                    change = True
                    break

                # once we can't change and still not satisfy `F` condition, directly return 
                if not change: return ""

        return "".join(word)

File: Greedy/3474. Lexicographically Smallest Generated String/solution.cpp
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    string generateString(string str1, string str2) {
        int n = str1.size(), m = str2.size();
        vector<char> word(n + m - 1);
        vector<bool> T(n + m - 1);

        for (int i=0; i<n; i++) {
            if (str1[i] == 'T') {
                for (int j=0; j<m; j++) {
                    if (word[i+j] && word[i+j] != str2[j]) return "";
                    word[i+j] = str2[j];
                    T[i+j] = true;
                }
            }
        }

        for (int i=0; i<n+m-1; i++) {
            if (!word[i]) {
                word[i] = 'a';
            }
        }

        for (int i=0; i<n; i++) {
            if (str1[i] == 'F') {
                bool diff = false;
                for (int j=0; j<m; j++) {
                    if (word[i+j] != str2[j]) {
                        diff = true;
                        break;
                    }
                }
                if (diff) continue;

                bool change = false;
                for (int j=m-1; j>=0; j--) {
                    if (T[i+j]) continue;

                    // Why changing to 'b' is enough
                    // Since str2 consists of lowercase letters, if a substring currently matches str2 and we change one character that isn't fixed by a 'T', we can always make it not match.

                    // If str2[j] was 'a', we change it to 'b'.

                    // If str2[j] was anything else, our default 'a' would have already made it different.

                    // Therefore, the only reason an 'F' substring matches str2 is if str2 itself has a specific pattern that matches our 'T's and 'a's. A single flip to 'b' (or the next available char) breaks the match.
                    if (word[i+j] == 'a') {
                        word[i+j] = 'b';
                    }

                    change = true;
                    break;
                }

                if (!change) {
                    return "";
                }
            }
        }

        string res(word.begin(), word.end());
        return res;
    }
};